wgsl/index.bs

```diff
@@ -737,22 +737,42 @@ the identifier will denote the object of the declaration appearing closest to
 that use.
 We say the identifier use <dfn noexport>resolves</dfn> to that declaration.
 
-Note: A declaration always precedes its identifier's scope.
-Therefore, the nearest in scope declaration of an identifier always precedes the
-use of the identifier.
+There are multiple levels of scoping depending on how and where things are
+declared.
+
+When an identifier is used, it must be in scope for some declaration, or as part of a directive.
+
+A declaration is at <dfn noexport>module scope</dfn> if the declaration appears outside
+the text of any other declaration.
+Module scope declarations are [=in scope=] of the entire program.
+That is, declarations at module scope may be referenced by statements and
+expressions that precede that declaration textually.
+
+It is a [=shader-creation error=] if any module scope declaration is recursive.
+That is, there must be no cycles among the module scope declarations.
+
+Note: The [=function body=] is part of the [=function declaration=], thus
+functions may not be recursive.
+
+Note: Use of a non-[=module scope=] identifier must succeed the declaration of
+that identifier in the text.
+This is not true, however, for [=module scope=] declarations, which may be
+referenced out of order in the text.
+
+Note: Only a [=function declaration=] can contain other declarations.
 
 <div class='example' heading='Valid and invalid declarations'>
   <xmp>
     // Invalid, cannot reuse built-in function names.
     var<private> modf: f32 = 0.0;
 
-    // Valid, foo_1 is in scope until the end of the program.
+    // Valid, foo_1 is in scope of the entire program.
```

Comment:
Should say "foo_1 is in scope for the entire program".  or "The scope of f_1 is the entire program"

Similarly for the cases below.